File: back/src/main/java/com/funstat/store/GenericDaoImpl.java
package com.funstat.store;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.funstat.domain.Ohlc;
import com.google.common.collect.Lists;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
import org.springframework.jdbc.datasource.DataSourceTransactionManager;
import org.springframework.transaction.support.TransactionTemplate;
import org.sqlite.SQLiteDataSource;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.stream.Collectors;

public class GenericDaoImpl implements GenericDao {

    private final DataSourceTransactionManager manager;
    final SQLiteDataSource ds;

    public GenericDaoImpl(SQLiteDataSource ds) {
        this.ds = ds;
        this.manager = new DataSourceTransactionManager(ds);
    }

    private HashMap<String, Object> getParamsMap(String code) {
        HashMap<String, Object> params = new HashMap<>();
        params.put("CODE", code);
        return params;
    }

    String write(Object obj) {
        ObjectMapper mapper = new ObjectMapper();
        ByteArrayOutputStream str = new ByteArrayOutputStream();
        try {
            mapper.writer().writeValue(str, obj);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return new String(str.toByteArray());

    }

    <T> T deser(String str, Class<T> clazz) {
        ObjectMapper mapper = new ObjectMapper();
        try {
            return mapper.readValue(str, clazz);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public synchronized <T> void saveGeneric(String type, List<T> obj, Function<T, String> keyMapper) {

        ensureExistGeneric(type);

        System.out.println("inserting " + obj.size() + " into " + type);

        HashMap[] data = obj.stream().map(ob -> {
            return new HashMap() {
                {
                    put("KEY", keyMapper.apply(ob));
                    put("JSON", write(ob));
                }
            };
        }).toArray(i -> new HashMap[i]);

        saveInTransaction("insert or replace into " + type + " (key,json) values (:KEY,:JSON)", data);

    }

    private void ensureExistGeneric(String type) {
        new JdbcTemplate(ds).execute("create table if not exists " + type + " (json varchar, key varchar primary key )");
    }


    private void saveInTransaction(String sql, Map[] data) {
        TransactionTemplate template = new TransactionTemplate(manager);
        template.execute(status -> {
            new NamedParameterJdbcTemplate(ds).batchUpdate(sql, data);
            return null;
        });
    }

    @Override
    public synchronized <T> List<T> readGeneric(String tableName, Class<T> clazz) {
        ensureExistGeneric(tableName);
        return new JdbcTemplate(ds).query("select * from " + tableName, (rs, rowNum) -> {
            return deser(rs.getString("json"), clazz);
        });
    }
}

File: back/src/main/java/com/funstat/store/GenericDaoImpl.kt
package com.funstat.store

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.KotlinModule

import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate
import org.springframework.jdbc.datasource.DataSourceTransactionManager
import org.springframework.transaction.support.TransactionTemplate
import org.sqlite.SQLiteDataSource
import java.io.ByteArrayOutputStream
import java.io.IOException

class GenericDaoImpl(internal val ds: SQLiteDataSource) : GenericDao {

    private val manager = DataSourceTransactionManager(ds)

    val mapper = ObjectMapper().registerModule(KotlinModule())

    fun write(obj: Any): String {
        val str = ByteArrayOutputStream()
        try {
            mapper.writer().writeValue(str, obj)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        return String(str.toByteArray())
    }

    fun <T> deser(str: String, clazz: Class<T>): T {
        try {
            return mapper.readValue(str, clazz)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }

    }

    @Synchronized
    override fun <T> saveGeneric(type: String, obj: List<T>, keyMapper: (T)->String) {

        ensureExistGeneric(type)

        println("inserting " + obj.size + " into " + type)

        val data = obj.map { ob ->
            mapOf("KEY" to keyMapper(ob), "JSON" to write(ob!!))
        }.toTypedArray()


        saveInTransaction("insert or replace into $type (key,json) values (:KEY,:JSON)", data as Array<Map<String,Any>>)

    }

    private fun ensureExistGeneric(type: String) {
        JdbcTemplate(ds).execute("create table if not exists $type (json varchar, key varchar primary key )")
    }


    private fun saveInTransaction(sql: String, data: Array<Map<String,Any>>) {
        val template = TransactionTemplate(manager)
        template.execute<Any> { status ->NamedParameterJdbcTemplate(ds).batchUpdate(sql, data)}
    }

    @Synchronized
    override fun <T> readGeneric(tableName: String, clazz: Class<T>): List<T> {
        ensureExistGeneric(tableName)
        return JdbcTemplate(ds).query("select * from $tableName") { rs, rowNum -> deser(rs.getString("json"), clazz) }
    }
}
